clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -237,12 +239,18 @@ public void close() {
      * @param send The request to send
      */
     public void send(Send send) {
-        KafkaChannel channel = channelOrFail(send.destination());
-        try {
-            channel.setSend(send);
-        } catch (CancelledKeyException e) {
-            this.failedSends.add(send.destination());
-            close(channel);
+        String connectionId = send.destination();
+        KafkaChannel closingChannel = closingChannels.remove(connectionId);
+        if (closingChannel != null)
+            doClose(closingChannel, true);
```

Comment:
I have moved the handling of close to `clear()`, so `closingChannels` is updated only in `clear()`. Can you check if it is any better? Thank you.